File: app/lib/src/models/ride_request.dart
import 'ride_group.dart';
import 'zone.dart';

class RideRequest {
  const RideRequest({
    required this.id,
    required this.status,
    required this.terminal,
    required this.zoneId,
    required this.windowStart,
    required this.windowEnd,
    required this.seats,
    required this.luggageCount,
    required this.genderPreference,
    this.zone,
    this.dropLandmark,
    this.flightNumber,
    this.note,
    this.rideGroupId,
    this.group,
  });

  factory RideRequest.fromJson(Map<String, dynamic> json) => RideRequest(
        id: json['id'] as int,
        status: json['status'] as String? ?? 'open',
        terminal: json['terminal'] as String? ?? 'T2',
        zoneId: json['zone_id'] as int? ?? 0,
        windowStart: DateTime.parse(json['window_start'] as String).toLocal(),
        windowEnd: DateTime.parse(json['window_end'] as String).toLocal(),
        seats: json['seats'] as int? ?? 1,
        luggageCount: json['luggage_count'] as int? ?? 1,
        genderPreference: json['gender_preference'] as String? ?? 'any',
        dropLandmark: json['drop_landmark'] as String?,
        flightNumber: json['flight_number'] as String?,
        note: json['note'] as String?,
        rideGroupId: json['ride_group_id'] as int?,
        zone: json['zone'] is Map<String, dynamic> ? Zone.fromJson(json['zone'] as Map<String, dynamic>) : null,
        group: json['group'] is Map<String, dynamic> ? RideGroup.fromJson(json['group'] as Map<String, dynamic>) : null,
      );

  final int id;
  final String status;
  final String terminal;
  final int zoneId;
  final DateTime windowStart;
  final DateTime windowEnd;
  final int seats;
  final int luggageCount;
  final String genderPreference;
  final Zone? zone;
  final String? dropLandmark;
  final String? flightNumber;
  final String? note;
  final int? rideGroupId;
  final RideGroup? group;

  bool get isOpen => status == 'open';

  bool get isMatched => status == 'matched';

  bool get isWomenOnly => genderPreference == 'women_only';
}

File: api/app/Models/RideRequest.php
<?php

namespace App\Models;

use App\Enums\CabService;
use App\Enums\GenderPolicy;
use App\Enums\RideRequestStatus;
use Carbon\CarbonInterface;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

#[Fillable([
    'user_id', 'airport_code', 'terminal', 'zone_id', 'drop_landmark', 'flight_number',
    'window_start', 'window_end', 'seats', 'luggage_count', 'gender_preference', 'note',
    'quoted_fare', 'cab_service', 'meeting_point',
])]
class RideRequest extends Model
{
    use HasFactory;

    /**
     * Mirrors the column defaults so a freshly created model reads the same in
     * memory as it does after a round trip to the database.
     */
    protected $attributes = [
        'airport_code' => 'BLR',
        'seats' => 1,
        'luggage_count' => 1,
        'gender_preference' => 'any',
        'status' => 'open',
    ];

    protected function casts(): array
    {
        return [
            'window_start' => 'datetime',
            'window_end' => 'datetime',
            'matched_at' => 'datetime',
            'cancelled_at' => 'datetime',
            'status' => RideRequestStatus::class,
            'gender_preference' => GenderPolicy::class,
            'cab_service' => CabService::class,
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function zone(): BelongsTo
    {
        return $this->belongsTo(Zone::class);
    }

    public function group(): BelongsTo
    {
        return $this->belongsTo(RideGroup::class, 'ride_group_id');
    }

    public function scopeOpen(Builder $query): Builder
    {
        return $query->where('status', RideRequestStatus::Open);
    }

    /**
     * Requests whose departure window overlaps [$start, $end] at all. The
     * minimum-overlap threshold is applied in MatchingService, not here.
     */
    public function scopeOverlapping(Builder $query, CarbonInterface $start, CarbonInterface $end): Builder
    {
        return $query->where('window_start', '<', $end)->where('window_end', '>', $start);
    }

    /**
     * Minutes this request's window shares with another window. Zero or less
     * means they never coincide.
     */
    public function overlapMinutesWith(CarbonInterface $start, CarbonInterface $end): int
    {
        $from = $this->window_start->max($start);
        $to = $this->window_end->min($end);

        return $to->greaterThan($from) ? (int) $from->diffInMinutes($to) : 0;
    }

    public function isOpen(): bool
    {
        return $this->status === RideRequestStatus::Open;
    }
}
